obs-ffmpeg: Add support for HLS ingestion

mv2731 opened the pull request:
Add and register an obs_output_info struct called
ffmpeg_hls_muxer, and initialize the struct with its
own start function. Within this start function, set a
user-agent setting so that the underlying HTTP request
made by ffmpeg will have a User-Agent header. Determine
whether the stream is hls to guess the output format in
ffmpeg-mux/ffmpeg-mux.c.

mv2731 committed:
obs-ffmpeg: Add support for HLS ingestion

Add and register an obs_output_info struct called
ffmpeg_hls_muxer, and initialize the struct with its
own start function. Within this start function, set a
user-agent setting so that the underlying HTTP request
made by ffmpeg will have a User-Agent header. Determine
whether the stream is hls to guess the output format in
ffmpeg-mux/ffmpeg-mux.c and use that in init functions.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -66,6 +66,8 @@ struct ffmpeg_muxer {
 
 	bool is_network;
+
+	struct dstr settings;
```
settings -> muxer_settings, and move it up after struct dstr path.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -211,19 +219,26 @@ static void log_muxer_params(struct ffmpeg_muxer *stream, const char *settings)
 	dstr_catf(cmd, "\"%s\" ", mux.array ? mux.array : "");
 
 	dstr_free(&mux);
+	dstr_free(&stream->settings);
```
Call dstr_free in ffmpeg_mux_destroy.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +369,51 @@ static bool ffmpeg_mux_start(void *data)
+	const char *path_str;
+	const char *stream_key;
+	struct dstr path = {0};
+	struct dstr settings = {0};
```
settings variable is unnecessary.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +369,51 @@ static bool ffmpeg_mux_start(void *data)
+	stream_key = obs_service_get_key(service);
+	dstr_copy(&path, path_str);
+	dstr_replace(&path, "{stream_key}", stream_key);
+	dstr_catf(&settings, "http_user_agent=libobs/%s", OBS_VERSION);
```
You can concatenate the string directly to stream->settings.

ushadow reviewed (changes requested):
Please see comments.

mv2731 committed:
obs-ffmpeg: Change settings name to muxer_settings

Change struct dstr settings to struct dstr muxer_settings
and then refactor muxer_settings code to remove unnecessary lines.
Move free statement of muxer_settings into destroy function.
